### DISCUSSION TITLE
Why rustfmt removes curly braces ? And how to prevent it ?

### DISCUSSION THREAD

**Discussion by yguel** [rust-lang/rustfmt#5663] [Q&A]
I have 2 code pieces, one where rustfmt removes curly braces and one where it does not and it annoys me.
It may be of no consequences, but I still fill confused by it especially because the output seems less readable to me.

First example where rustfmt does not remove curly braces
```rust
use std::thread;
use std::time::Duration;

fn main() {
    let handle = thread::spawn(|| {
        for i in 1..10 {
            println!("hi number {} from the spawned thread!", i);
            thread::sleep(Duration::from_millis(1));
        }
    });
    for i in 1..5 {
        println!("hi number {} from the main thread!", i);
        thread::sleep(Duration::from_millis(1));
    }

    handle.join().unwrap();
}
```

Here is the second code before use of rustfmt:
```rust
use std::thread;
use std::time::Duration;

fn main() {
    let handle = thread::spawn(|| {
        loop {
            println!("hi from the spawned thread!");
            thread::sleep(Duration::from_millis(1));
        }
    });

    for i in 1..5 {
        println!("hi number {} from the main thread!", i);
        thread::sleep(Duration::from_millis(1));
    }

    handle.join().unwrap();
}
```

And now the code after rustfmt pass (the loop statement changes):
```rust
use std::thread;
use std::time::Duration;

fn main() {
    let handle = thread::spawn(|| loop {
        println!("hi from the spawned thread!");
        thread::sleep(Duration::from_millis(1));
    });

    for i in 1..5 {
        println!("hi number {} from the main thread!", i);
        thread::sleep(Duration::from_millis(1));
    }

    handle.join().unwrap();
}
```

I like the fact that the closure is presented in its own block even if its only a loop block. 
Furthermore, I very hate that a tool is removing some code even if it is unnecessary braces. I would prefer suggestions for that.
I would love to configure rustfmt to not remove code at all and just work on indentation. Is there an option for that ?

**Comment by wellkilo** [comment]
The transformation is intentional: the closure body is a block containing exactly one expression (the loop), so rustfmt can remove the outer closure block. These two closure bodies are semantically equivalent:

    || { loop { /* ... */ } }
    || loop { /* ... */ }

Your first example stays braced because that closure block contains more than one statement: the for loop and handle.join().

There is a setting for the specific preference you want:

    # rustfmt.toml
    unstable_features = true
    force_multiline_blocks = true

Run it with nightly rustfmt:

    cargo +nightly fmt

force_multiline_blocks forces closure and match-arm bodies to remain wrapped in a multiline block. It is currently an unstable rustfmt option, so stable rustfmt will not honor it. See the [configuration reference](https://github.com/rust-lang/rustfmt/blob/9e92f7b48cf49412c647eb6b20b6e4315481a23f/Configurations.md#force_multiline_blocks) and the [closure-body decision](https://github.com/rust-lang/rustfmt/blob/9e92f7b48cf49412c647eb6b20b6e4315481a23f/src/closures.rs#L455-L475).

There is no general indentation-only mode that promises never to rewrite equivalent syntax. If you cannot use nightly, the narrow escape hatch is #[rustfmt::skip] on the containing item, but that disables all formatting for the item and is usually less desirable than accepting the canonical form.

If this preserves the closure layout you wanted, please mark the reply as accepted.

**Comment by yguel** [comment]
> Your first example stays braced because that closure block contains more than one statement: the for loop and handle.join().

I completely miss your point there, I do not see that there is a different number of statements in the 2 examples in the closure.

For me the ``handle.join`` is in the main not in the closure definition.
Hence the different behaviour is really troubling.


The setting for the preference I want looks good. I will try it in multiple cases to tell you if this reply can be accepted.
Nevertheless, the difference in the way those 2 pieces of code are handled is not logical:  the example provided is quite explicit about it. This, alone, justify why it should not be happening IMHO.
